rfcs/20200306-single-client-parameter-server.md

```diff
@@ -0,0 +1,517 @@
+for x in enumerate(distributed_iter):
+  loss = strategy.schedule(replica_fn, x, y)
+```
+
+
+If we do the same thing with the `strategy.schedule` API, there are several challenges.
+
+The first challenge is we don’t know which worker the `get_next` should return to since where the `replica_fn` will be executed will be decided later. Some later-binding mechanism can be explored.
+
+The second challenge is calling `get_next` on an iterator is synchronous. This means that the training loop is not truly asynchronous. It is tricky to make `get_next` asynchronous because the client doesn’t know how many items will be in the iterator and thus doesn’t know how many functions to schedule. 
+
+
+##### Alternative: passing iterators to `strategy.schedule`
+
+The following training loop is less consistent with other `tf.distribute` examples but is easier to implement in the short term. It requires users to explicitly set a number of steps.
+
+
+```python
+# … omitted
+with strategy.scope():
+  # … omitted
+  distributed_iter = iter(distributed_dataset)
+  for i in range(total_steps):
+    strategy.schedule(replica_fn, args=(distributed_iter,))
+# … omitted
+```
+
+
+**We will start with this kind of training loop in our first version.**
+
+
+### Fault Tolerance
+
+
+#### Task Failure
+
+
+##### Worker failure
+
+
+###### When scheduling
+
+When a worker fails, our training will continue without this failed worker. Functions scheduled on a failed worker will be rescheduled on other workers. 
+
+For functions that bound to a specific worker, e.g. resource creation function, they will be queued until the worker is back.
+
+When the failed worker is back, we will update the cluster configuration with `context.update_server_def` which would also reset all the states. After resources on the restarted worker are built, we can resume scheduling functions on the worker. 
+
+
+###### When materialing a `Future` object
+
+It is possible that a function is executed but its corresponding worker fails when users try to consume its output. In this case, we will give users a `None` value and set an error in the `Future` object.
+
+We can mitigate the problem by eagerly materializing function outputs when they are passed to `local_results`. 
+
+We can explore mechanisms to recover these objects in the future. In the short-term, users can choose to write the results to variables on parameter servers, just like a Keras metric.
+
+
+##### Parameter server failure
+
+When a parameter server fails, the error will be propagated to the client via workers. Since the latest values of variables on the failed parameter servers are gone, there is no way for the client to recover them. Therefore the training will pause until the failed parameter server is back. The client then needs to clean up other variables on other parameter servers, rebuild all the variables and load variable values from a checkpoint. To trigger this process, the simplest method is to restart the client as well.
```

Comment:
Can you give a more concrete example of how training/the user will recover here? I have two parameter servers with ten workers each. What happens to 22 (or 23? Is the single-client one of the workers?) nodes if one of the param servers goes down?

Reply:
I believe currently all ps's and workers (and client) are considered the same cluster, so if any of the ps's fails, the cluster restarts (yuefengz@ to correct me if wrong).

Reply:
PS failure: client restart and wait until the failing ps is back.